# Running EPACTS in DNAnexus on hg38 data

Here is a quick tuturial on running EPACTS in DNAnexus. We use custom shell scripts, called from the *Swiss Army Knife* command line to run EPACTS commands (just like you would locally) together with a docker image containing all of the EPACTS code. We'll use Topmed freeze5b data from the Amish cohort in this example.

## Getting Started

First, you'll need to login to DNAnexus and create a new folder for your EPACTS scripts and data. This should be in some DNANexus project where you would like to run the analysis.

In this tutorial, we'd like to generate and use a kinship matrix in a single variant association analysis. We'll do this by running Epacts/Emmax on our input genotype and phenotype files.

We will be using some sample data but you should be free to use whatever genotype and phenotype data you would like. For this tutorial, you will need:
1. [bgzipped](http://www.htslib.org/doc/tabix.html) genotype file in vcf format
2. [tabix](http://www.htslib.org/doc/tabix.html) index file of the vcf
3. phenotype file in [Epacts format](https://genome.sph.umich.edu/wiki/EPACTS#PED_file_for_Phenotypes_and_Covariates)

These files must be uploaded into your project folder with the DNANexus add data tool. It should look like this:

![adding data to DNANexus](https://user-images.githubusercontent.com/30277822/34653670-5915b8b6-f3bd-11e7-84e6-31452fc2ed86.png)

## Writing scripts for DNAnexus

Running EPACTS on DNAnexus boils down to copying your normal command line input to a shell script. The script is then interpreted within a docker image, and the EPACTS commands executed. For more information on running EPACTS from the commend line, see their [documentation](https://genome.sph.umich.edu/wiki/EPACTS). From here, we'll assume that you know the basics of using EPACTS. 

The relevant Epacts commands are:

```bash
touch demo.vcf.bgz.tbi
$runepacts make-kin --vcf demo.vcf.bgz --min-maf 0.01 --out demo.kinf --run 2 --buildver hg38
$runepacts single --vcf demo.vcf.bgz --ped epacts_demo.ped --min-maf 0.001 --kin demo.kinf --pheno FG --cov SEX --cov AGE --test q.emmax --out demo_sv --run 2 --buildver hg38
```

For hg38 data, we need to include the input parameter `--buildver hg38`. Otherwise, Epacts defualts to hg19.

To run this on DNANexus through the Swiss Army Knife app, we'll need to save it as a shell script and upload it to the project folder.

## Running the script from the GUI

1. From the project folder folder, we'll need to create a new workflow.

![untitled workflow](https://user-images.githubusercontent.com/30277822/34653699-9f5bb212-f3bd-11e7-982d-a9ea9abeed95.png)

2. Next we need to add the Swiss Army Knife app by clicking on `Add a Step` and searching for the app.

3. Click `inputs` near the left side of page to select input files from the project folder. These should be your vcf, index file, phenotype file, and shell script.

4. Next, click `Swiss Army Knife` in black near the center of the screen. 

5. We'll need to modify the `Command line` and `Optional Docker image identifier` fields. For our shell script `demo.sh`, it should look like this:

![filled workflow](https://user-images.githubusercontent.com/30277822/34653708-bbdcd9b6-f3bd-11e7-9b15-1badd095bce9.png)

7. Under `Optional Docker image identifier`, make sure we have `tmajarian/epacts-rmkl:latest`.

8. Click `Save` then `Start Analysis...` in green to run the script.

![submit job](https://user-images.githubusercontent.com/30277822/34653712-d7952f50-f3bd-11e7-84e2-4a4d32cf94bc.png)
