# Running EPACTS in DNAnexus on hg38 data

Here is a quick tuturial on running EPACTS in DNAnexus. We use custom shell scripts, called from the *Swiss Army Knife* App command line to run EPACTS (just like you would locally) together with a docker image containing all of the EPACTS code. In this tutorial, we'll use genotype data in hg38 coordinates.

## Getting Started

First, you'll need to login to DNAnexus and create a new folder for your EPACTS scripts and data. This should be in some DNANexus project where you would like to run the analysis.

In this tutorial, we'd like to generate and use a kinship matrix in a single variant association analysis. We'll do this by running Epacts/Emmax on our input genotype and phenotype files.

The data that you will need is:
1. [bgzipped](http://www.htslib.org/doc/tabix.html) genotype file in vcf format
2. [tabix](http://www.htslib.org/doc/tabix.html) index file of the vcf
3. phenotype file in [Epacts format](https://genome.sph.umich.edu/wiki/EPACTS#PED_file_for_Phenotypes_and_Covariates)

These files must be uploaded into your project folder with the DNANexus *add data* tool. Navigate to your new project folder and add data. It should look like this:

![adding data to DNANexus](https://user-images.githubusercontent.com/30277822/34653670-5915b8b6-f3bd-11e7-84e6-31452fc2ed86.png)


## Writing scripts for DNAnexus

Now that the data is in place, we'll need to write the actual EPACTS commands. Running EPACTS on DNAnexus boils down to copying your normal command line input to a shell script. The script is then called within a docker image, and the EPACTS commands executed. For more information on running EPACTS from the commend line, see their [documentation](https://genome.sph.umich.edu/wiki/EPACTS). From here, we'll assume that you know the basics of using EPACTS. 

First, we need to take care of an EPACTS requirement -- that the index file is 'newer' than the VCF. We can accomplish this by running:

```bash
touch demo.vcf.bgz.tbi
```

Next, we need to generate a kinship matrix. We'll also choose some optional inputs: a minimum allele frequency of 1% for included variants (--min-maf), an output file name (--out), and the number of parallel processes (--run). To use hg38 data, we also need to input the *--buildver hg38* flag. Notice here that we can use the global variable `$runepacts` instead of the absolute path to the EPACTS executable. This is built into the docker image for convenience and will only work from inside the container. The command is then:

```bash
$runepacts make-kin --vcf demo.vcf.bgz --min-maf 0.01 --out demo.kinf --run 2 --buildver hg38
```

Finally, to run a single variant association analysis, we'll use the generated kinship matrix, genotype data, and phenotype data. We'll also need to set some input parameters: minimum allele frequency (--min-maf), the outcome we're testing (--pheno), covariates (--cov), an output file prefix (--out), the number of parallel jobs (--run), and the genome build version (--buildver)

```bash
$runepacts single --vcf demo.vcf.bgz --ped epacts_demo.ped --min-maf 0.001 --kin demo.kinf --pheno FG --cov SEX --cov AGE --test q.emmax --out demo_sv --run 2 --buildver hg38
```

To run this on DNANexus through the Swiss Army Knife app, we'll need to put all of these commands together in a shell script and upload it to the project folder. The final shell script should be:

```bash
touch demo.vcf.bgz.tbi
$runepacts make-kin --vcf demo.vcf.bgz --min-maf 0.01 --out demo.kinf --run 2 --buildver hg38
$runepacts single --vcf demo.vcf.bgz --ped epacts_demo.ped --min-maf 0.001 --kin demo.kinf --pheno FG --cov SEX --cov AGE --test q.emmax --out demo_sv --run 2 --buildver hg38
```

## Running the analysis from the GUI

1. From the project folder folder, we'll need to create a new workflow by clicking: ![new workflow](https://user-images.githubusercontent.com/30277822/34676161-efcdf39e-f459-11e7-8929-013612c06ab7.png).

2. Next add the Swiss Army Knife app by clicking on ![Add a Step](https://user-images.githubusercontent.com/30277822/34676208-1785c0b0-f45a-11e7-9458-2ef843b50010.png) and searching for the app.

![search](https://user-images.githubusercontent.com/30277822/34676242-31290aa4-f45a-11e7-9bcc-d92289876c6d.png)

3. Click ![input files [array]](https://user-images.githubusercontent.com/30277822/34676262-3f08380c-f45a-11e7-9b5f-f908b88f1537.png) near the left side of page to select input files from the project folder. These should be your vcf, index file, phenotype file, and shell script.

4. Next, click ![Swiss Army Knife](https://user-images.githubusercontent.com/30277822/34676291-5740cd1c-f45a-11e7-9ddb-fd805fc04f7f.png) near the center of the screen. 

5. We'll need to modify the `Command line` and `Optional Docker image identifier` fields. For our shell script `demo.sh`, it should look like this:

![command line docker image](https://user-images.githubusercontent.com/30277822/34676327-73914d84-f45a-11e7-9e7b-d558244a9d9b.png)

7. Under `Optional Docker image identifier`, make sure we have `tmajarian/epacts-rmkl:latest`.

8. Click *Save* then ![Start Analysis...](https://user-images.githubusercontent.com/30277822/34676354-8585c628-f45a-11e7-91ca-8ad1f41c2135.png) to run the script.
